### PR TITLE
[serve] Prefer routing to replicas on the same node

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

This change modifies the router to prefer scheduling requests to replicas on the same node.

The procedure is simple: in the first iteration of replica selection, the router considers only replicas on the same node (if any are available). In subsequent iterations, it falls back to considering all replicas.

The reasoning for this procedure rather than attempting _all_ replicas on the same node first is that if the chosen replicas have their queues full, it's likely that the system is under heavy load and the other replicas will be full as well.

Note: This change is guarded behind a feature flag and turned off by default.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by shrekris-anyscale — COMMENTED** [review]
This approach makes sense overall.

File: python/ray/serve/handle.py [hunk 1]
```diff
@@ -160,6 +160,7 @@ def _get_or_create_router(self) -> Router:
             self._router = Router(
                 serve.context.get_global_client()._controller,
                 self.deployment_name,
+                ray.get_runtime_context().get_node_id(),
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/handle.py:163]: Nope, returns the node_id driver is on:
```
>>> import ray
>>> ray.init()
2023-08-10 11:50:38,781 INFO worker.py:1431 -- Connecting to existing Ray cluster at address: 127.0.0.1:6379...
2023-08-10 11:50:38,814 INFO worker.py:1612 -- Connected to Ray cluster. View the dashboard at http://127.0.0.1:8265
RayContext(dashboard_url='127.0.0.1:8265', python_version='3.8.13', ray_version='3.0.0.dev0', ray_commit='{{RAY_COMMIT_SHA}}', protocol_version=None)
>>> ray.get_runtime_context().get_node_id()
'156e1b14ad51e4d8e8442690b6f6e3eb0af2d66316b53575e0051ad8'
```

**Comment by edoakes** [comment]
Going to re-run @shrekris-anyscale's scalability benchmarks against this and see how it performs

File: python/ray/serve/_private/router.py [hunk 2]
```diff
@@ -296,18 +300,22 @@ def __init__(
         self,
         event_loop: asyncio.AbstractEventLoop,
         deployment_name: str,
+        self_node_id: Optional[str] = None,
     ):
         self._loop = event_loop
         self._deployment_name = deployment_name
+        self._self_node_id = self_node_id
```
Comment (zcin) [inline python/ray/serve/_private/router.py:307]: Nit: can this be evaluated in `PowerOfTwoChoicesReplicaScheduler.__init__` on the spot instead of passed in when creating the router?

**Review by zcin — COMMENTED** [review]
Curious: If we have an application with a driver + downstream deployment (e.g. DAGDriver + resnet deployment), and there is only one DAGDriver replica but many resnet replicas, will this cause imbalance in routing? Since this might cause the DAGDriver to keep sending requests to the resnet replicas on its own node until it maxes out their `max_concurrent_queries`, then send requests to the replicas on the other nodes?

File: python/ray/serve/_private/router.py (hunk 2 above)
Comment (zcin) [reply python/ray/serve/_private/router.py:307]: Gotcha, thanks a lot for the explanation!

**Comment by edoakes** [comment]
@zcin regarding your top-level comment, yes that is the case. In some cases this would be a performance regression (e.g., if the `max_concurrent_queries` is tuned to be high and there's a single driver).

This change relies much more on `max_concurrent_queries` being tuned properly to avoid such issues. One option to mitigate the blast radius is to make this behavior optional and opt-in to start.

**Comment by zcin** [comment]
@edoakes Yeah, makes sense. It'd be nice if we can reduce the default `max_concurrent_queries` in this case, since by default it's pretty high and users probably rarely think to tune it.

Do you think it would be reasonable to not limit the candidate replicas, but instead pick two random replicas then apply node preference when deciding on one out of the two replicas? I feel like that could remove the issue of a driver overloading replicas on its own node. E.g, after choosing two random replicas
1. First, prefer replica on same node
2. Second, prefer replica with smaller queue

Or is it possible to guard your current approach behind a flag in `Router`, and only turn on that flag for HTTP proxies?

File: python/ray/serve/_private/router.py [hunk 3]
```diff
@@ -296,18 +300,22 @@ def __init__(
...
 
         # Current replicas available to be scheduled.
         # Updated via `update_replicas`.
         self._replica_id_set: Set[str] = set()
         self._replicas: Dict[str, ReplicaWrapper] = {}
         self._replicas_updated_event = make_asyncio_event_version_compat(event_loop)
+        self._replica_ids_colocated_on_same_node: Set[str] = set()
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/router.py:314]: nit: Can we shorten this to just `_colocated_replica_ids` and call out in a comment that colocation assumes same node

File: python/ray/serve/tests/test_handle.py [hunk 4]
```diff
@@ -385,6 +386,60 @@ def echo(name: str):
     ), handle2._router._replica_scheduler
 
 
+def test_handle_prefers_replicas_on_same_node(ray_cluster):
```
Comment (alexeykudinkin) [inline python/ray/serve/tests/test_handle.py:389]: Should we also test the other branch (when there are no colocated replicas)?

**Review by alexeykudinkin — APPROVED** [review]

**Comment by alexeykudinkin** [comment]
> 1. First, prefer replica on same node
> 2. Second, prefer replica with smaller queue

@zcin did i understand your proposal correctly that you're suggesting to pick 1 colocated and 1 random replica and the use standard Po2C?

I think that's an interesting idea and we can experiment with it, but i'd rather keep things simple from the get-go and additional functionality if needed. 

Primary goal of local routing is to minimize spilling out-of-memory RPC as much as possible (ie localizing the traffic to a node) for as long as it's able to handle.

> That would be less of a performance boost for HTTP Proxies, though, especially if there's a lot of nodes. Would it be useful to guard your current approach behind a flag in Router, and only turn on that flag for HTTP proxies?

HTTPProxy should definitely *not* be spilling traffic to other nodes while their own can handle it (otherwise it becomes double-stacked load-balancing in some setups which doesn't really make sense).

That being said we should definitely keep it configurable though so that we don't try to fit all cases into same t-shirt size.

File: python/ray/serve/_private/router.py [hunk 5]
```diff
@@ -568,8 +587,17 @@ async def choose_two_replicas_with_backoff(
...
                         )
                     )
+            elif (
+                self._locality_routing
+                and backoff_index == 0
+                and len(self._colocated_replica_ids) > 0
+            ):
+                # Attempt to schedule requests to a replica on the same node in the
+                # first iteration only.
+                candidate_replica_ids = self._colocated_replica_ids
```
Comment (zcin) [inline python/ray/serve/_private/router.py:597]: Guard the approach behind the flag `locality_routing`, which is off for handles by default, but turned on for HTTP proxies.

**Comment by zcin** [comment]
@alexeykudinkin @edoakes please take another look! 
`_locality_routing` is turned on for HTTP proxies but otherwise off by default. The flag is turned on for `test_handle_prefers_replicas_on_same_node`. Tests in `test_replica_scheduler.py` are also run with the flag both on and off.

**Comment by zcin** [comment]
After discussion with @sihanwang41 also added an env variable `RAY_SERVE_ENABLE_PROXY_LOCALITY_ROUTING` which can be used to turn off the locality routing for http proxies in case we need it.

**Comment by sihanwang41** [comment]
Hi @akshay-anyscale, I just talked with @shrekris-anyscale, we haven't done any benchmarking with bigger payloads size between same node and corss code, and it is near the branch cut, I propose to guard it with feature flag. (Or we only merge before we see a clear benchmarking improvements for specific workloads and no regression, which i think it is not likely to happen before 2.7). cc: @edoakes

File: python/ray/serve/_private/constants.py [hunk 6]
```diff
@@ -226,6 +226,12 @@
     os.environ.get("RAY_SERVE_ENABLE_NEW_HANDLE_API", "0") == "1"
 )
 
+# Feature flag to turn on locality routing for HTTP proxies.
+# This is currently ON BY DEFAULT.
+RAY_SERVE_ENABLE_PROXY_LOCALITY_ROUTING = (
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/constants.py:231]:
```suggestion
RAY_SERVE_PROXY_PREFER_LOCAL_ROUTING = (
```

File: python/ray/serve/_private/common.py [hunk 7]
```diff
@@ -393,6 +394,7 @@ def __post_init__(self):
                 [
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/common.py:394]: Can we also fix the separator here? Using " " is problematic (if your node_id is empty, your new hash will be indistinguishable from old one)

Comment (alexeykudinkin) [reply python/ray/serve/_private/common.py:394]: You can use a char that we know for sure wouldn't be used in either of these fields (":", ",", etc)

File: python/ray/serve/_private/proxy_router.py [hunk 8]
```diff
@@ -65,6 +66,7 @@ def update_routes(self, endpoints: Dict[EndpointTag, EndpointInfo]) -> None:
                         and not info.app_is_cross_language
                     ),
                     use_new_handle_api=True,
+                    _locality_routing=RAY_SERVE_ENABLE_PROXY_LOCALITY_ROUTING,
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/proxy_router.py:69]:
```suggestion
                    _prefer_local_routing=RAY_SERVE_ENABLE_PROXY_LOCALITY_ROUTING,
```

File: python/ray/serve/_private/router.py [hunk 9]
```diff
@@ -349,18 +353,25 @@ def __init__(
         self,
         event_loop: asyncio.AbstractEventLoop,
         deployment_id: DeploymentID,
+        locality_routing: bool = False,
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/router.py:356]:
```suggestion
        prefer_local: bool = False,
```

File: python/ray/serve/tests/test_replica_scheduler.py [hunk 10]
```diff
@@ -476,8 +499,47 @@ async def test_replica_responds_after_being_removed(pow_2_scheduler, fake_query)
     assert (await task) == r2
 
 
+@pytest.mark.asyncio
+@pytest.mark.parametrize("pow_2_scheduler", [True], indirect=True)
```
Comment (alexeykudinkin) [inline python/ray/serve/tests/test_replica_scheduler.py:503]: We can just add a default, to avoid needing to do that

File: python/ray/serve/_private/common.py (hunk 7 above)
Comment (zcin) [reply python/ray/serve/_private/common.py:394]: Synced offline, will keep in this PR

**Comment by zcin** [comment]
@alexeykudinkin I've addressed comments, PTAL!

**Comment by edoakes** [comment]
@zhe-thoughts tagged

**Comment by zhe-thoughts** [comment]
@edoakes There's conflict now

**Review by alexeykudinkin — APPROVED** [review]

**Comment by edoakes** [comment]
conflict fixed; tests running

**Comment by zhe-thoughts** [comment]
@edoakes please reapply `tests-ok` when rerun finishes

**Review by zhe-thoughts — APPROVED** [review]
